### PR TITLE
refactor(core): clean pending invoice 'markAsPaid' signals

### PR DESCRIPTION
**Pull request by vindard**
## Description

This is a refactor to clean up and clarify when we mark transactions as "processed" or "paid and processed" depending on the outcome of the "update pending invoices" function.

This was prompted by issues observed where invoices would sometimes be wrongly marked as paid.

### CODE REVIEW

**Review by bodymindarts — APPROVED** [review]

File: core/api/src/app/wallets/update-single-pending-invoice.ts [hunk 1]
```diff
@@ -305,30 +312,30 @@ const lockedUpdatePendingInvoiceSteps = async ({
...
 
+  // Idempotent settle invoice in lnd
   if (!isSettledInLnd) {
     const lndService = LndService()
     if (lndService instanceof Error) {
       logger.error("Unable to initialize LndService")
-      recordExceptionInCurrentSpan({ error: lndService })
       return ProcessPendingInvoiceResult.err(lndService)
     }
+    // Returns 'true' on re-runs if invoice is already settled in lnd
```
[the remark this answers is not in the document]
Comment (vindard) [reply core/api/src/app/wallets/update-single-pending-invoice.ts:330]: We do higher up I believe, but I just noticed that it's a `paid` check and not a `processingCompleted` check. Fixed here https://github.com/GaloyMoney/galoy/pull/4028/commits/52b1db184e950cf0f57354134be06f04263f80b7

**Review by bodymindarts — APPROVED** [review]

File: core/api/src/app/wallets/update-single-pending-invoice.ts [hunk 2]
```diff
@@ -173,7 +183,7 @@ const processPendingInvoice = async ({
     currency: WalletCurrency.Btc,
   })
   if (receivedBtc instanceof Error) {
-    return ProcessPendingInvoiceResult.paidWithError(receivedBtc)
+    return ProcessPendingInvoiceResult.processAsPaidWithError(receivedBtc)
```
Comment (dolcalmi) [inline core/api/src/app/wallets/update-single-pending-invoice.ts:186]: if we have a conversion error it should be marked as paid/processed?

File: core/api/src/app/wallets/update-single-pending-invoice.ts [hunk 3]
```diff
@@ -348,21 +363,31 @@ const lockedUpdatePendingInvoiceSteps = async ({
...
+      error: journal,
+      level: ErrorLevel.Critical,
+      attributes: {
+        ["error.actionRequired.message"]:
+          "Check that invoice exists and is settled in lnd, unmark paid (and processed) from ledger entries",
+        ["error.actionRequired.paymentHash"]: paymentHash,
+        ["error.actionRequired.pubkey]"]: walletInvoiceInsideLock.pubkey,
+      },
+    })
+    return ProcessPendingInvoiceResult.err(journal)
```
Comment (dolcalmi) [inline core/api/src/app/wallets/update-single-pending-invoice.ts:376]: I have my doubts about this.. what if the error is in our side and not in db? should we include validation against medici txs too?

**Review by dolcalmi — CHANGES_REQUESTED** [review]
just few questions.. also... I couldn't find the milisats tests did you remove them in the bats refactor?

File: core/api/src/app/wallets/update-single-pending-invoice.ts (hunk 2 above)
Comment (vindard) [reply core/api/src/app/wallets/update-single-pending-invoice.ts]: True, this is the same kind of error as the `checkedToSats` one just above it, a validation error from an invalid amount passed. I made a change to treat them both the same and to decline the invoice if an invalid amount somehow comes through the process https://github.com/GaloyMoney/galoy/pull/4028/commits/cdcfcd764263b83477fb6f59b707d69ec4571f1e

File: core/api/src/app/wallets/update-single-pending-invoice.ts (hunk 3 above)
Comment (vindard) [reply core/api/src/app/wallets/update-single-pending-invoice.ts:383]: Message changed inside this [`force-push`](https://github.com/GaloyMoney/galoy/compare/cdcfcd764263b83477fb6f59b707d69ec4571f1e..ee294b7bbe2a5c9a9a8b335576a15389f9de28c7)

**Review by dolcalmi — APPROVED** [review]

**Pull request merged**